File: typescript/word-count/word-count.ts
type WordCounter = Map<string, number>;

export default class Words {
  count(sentence: string) {
    return sentence
      .trim()
      .split(/\s+/)
      .reduce(this.buildWordCounter, new Map<string, number>());
  }

  buildWordCounter(wordCounter: WordCounter, word: string) {
    const lowerCasedWord = word.toLowerCase();
    const currentCounter = wordCounter.get(lowerCasedWord);
    const counter = currentCounter ? currentCounter + 1 : 1;

    return wordCounter.set(lowerCasedWord, counter);
  }
}

File: elixir/word-count/lib/word_count.ex
defmodule WordCount do
  def count(sentence) do
    sentence
    |> String.downcase
    |> split_by_special_chars
    |> build_counter
  end

  defp split_by_special_chars(string) do
    String.split(string, ~r/[^[:alnum:]-]/u, trim: true)    
  end

  defp build_counter(strings) do
    Enum.reduce(strings, %{}, &update_counter/2)
  end

  defp update_counter(string, counter) do
    Map.update(
      counter,
      string, 1,
      &increment_counter/1
    )
  end

  defp increment_counter(string_counter) do
    string_counter + 1
  end
end
